Reserving values for additional payload types

expanded-for-real opened the pull request:
Reserving values for complex types as described in - https://github.com/imprint-serde/imprint-java/issues/25

expanded-for-real committed:
Update README.md

Added types for https://github.com/imprint-serde/imprint-java/issues/25

expanded-for-real committed:
Update README.md

fix spacing issue

expanded-for-real committed:
Update README.md

fix heading size for payload type

agavra commented on README.md:
```diff
@@ -147,7 +147,13 @@ Each entry has the following format:
+|         0xC | `time`     | 8-byte int64 milliseconds since midnight               |
+|         0xD | `uuid`     | 16-byte binary representation                          |
+|         0xE | `decimal`  | varint scale + varint length + unscaled bytes          |
+|         0xF | `timestamp`| 8-byte int64 milliseconds since Unix epoch (UTC)      |
```
super nitpick but the perfectionist in me is bothered by the fact that `timestamp` and `time` are not one after the other 😆 if it isn't a big deal, it'd be nice to have them both and `date` in the same block

expanded-for-real replied:
Can do!

agavra commented on README.md:
```diff
@@ -147,7 +147,13 @@ Each entry has the following format:
+|         0xB | `date`     | 4-byte int32 days since Unix epoch                     |
+|         0xC | `time`     | 8-byte int64 milliseconds since midnight               |
+|         0xD | `uuid`     | 16-byte binary representation                          |
+|         0xE | `decimal`  | varint scale + varint length + unscaled bytes          |
```
decimal is the only one here that's non-standard, we should define exactly the format that we are using here and add a reference to a standard (or if you don't want to do that now we can leave it for a follow up) -- I know that nailing down the serialization format for decimal has been a big pain in protobuf for a while (see github.com/protocolbuffers/protobuf/issues/4406)

expanded-for-real replied:
Might need to think about this for a minute and get back to you. I guess they worked up some proposals way back but it also never went anywhere due to challenges and prioritization - https://github.com/protocolbuffers/protobuf/pull/7039. Our proposal, which I would call the naive one (since Java has a `BigDecimal(BigInteger unscaledVal, int scale)` constructor), is basically [here](https://github.com/protocolbuffers/protobuf/pull/7039/files#diff-fd1de4249953a8fe58ac3c5609c3fba52fdbb2e0643c4cd766650d243b7544d2R250). 

And it looks like they ultimately recommended just going as a String value. That would work and would be very easy but we also take a performance hit of de/serializing Strings. I'll ponder this a bit

expanded-for-real replied:
@agavra after researching this and thinking about it a bit more, and looking through protobuf's proposed options - I think the proposed format above is acceptable for something that has to be self-describing (Avro and SBE benefit from schema based designs obviously). It can be used to construct proper decimal types in languages that support them natively (Java, C#, Python) and I think it has decent interoperability with Avro (Avro -> Imprint is trivial at least). As far as standards go - we're pretty close to **IEEE 754** (we should have no problem being able to represent any IEEE 754 decimal with exact precision).

We can pull decimal's out and create an issue for opening it up to further discussion too if you think it's warranted.

agavra replied:
This looks good to me! I think it's OK for us to be opinionated here if it has decent inter-op with Avro. Let's go with it and if it proves to be a challenge we can revise it at 1.0

agavra commented on README.md:
```diff
@@ -147,7 +147,13 @@ Each entry has the following format:
 |         0xA | `row`      | Nested Imprint row (recursive joins)                   |
-|      10–127 | *reserved* | Future primitives / logical types                      |
+|         0xB | `date`     | 4-byte int32 days since Unix epoch                     |
+|         0xC | `time`     | 8-byte int64 milliseconds since midnight               |
```
Is this standard to encode this as 8 bytes? I was checking and there are 8.64e+7 milliseconds in a day and max int is ~4.2e+9 -- definitely should align with other systems, but that's odd to me.

expanded-for-real replied:
Good catch - I originally had this proposed as microseconds but didn't actually adjust the type. I double checked Avro and it's int32 there as well for milliseconds.

agavra reviewed (commented):
Thanks @expanded-for-real -- I feel like you're already the most knowledgeable contributor to Imprint after all your work on the Java format so it's great to have you stamp (or should I say fingerprint 😆) on the main project.

Left some comments inline

agavra commented:
Not sure why the bench CI/CD isn't working but this obviously doesn't effect that so I'll dig into that another time. We can ignore it for now.

expanded-for-real committed:
Update README.md

reorganize time types and fix int representation for timestamp from int64 to int32
